Improve demo store SEO & ld+json implementation

juanpprieto opened the pull request:
<!--
  How to write a good PR title:
  - Start with a verb, for example: Add, Delete, Improve, Fix…
  - Give as much context as necessary and as little as possible
-->

### WHY are these changes introduced?

Fixes first task of https://github.com/Shopify/hydrogen/issues/593

TODO: 
- Add collection seo + ldJson
- Add journal/blog seo + ldJson
- Add pages seo + ldJson

### WHAT is this pull request doing?

Done
- Moved `root.tsx` `seo` to `loader` pattern and added `Organization` ldJson
- added `ldJson` to `products/$productHandle` based on [Dawn Theme](https://theme-dawn-demo.myshopify.com/).
- added a couple of wip `analyticsPayload` and `seoPayload` helpers to both templates to make them improve the DX and cosistency.

Issues:
- ldJson currently does not allow for multiple tags to be defined at root and route level. This is a pattern in Dawn so we might need to change the ldJson API to allow for this. We are discussing with Paul Shapiro.

Suggestion: 

Instead of having a hook `useAnalytics` should we instead have a `<Analytics />` component to better align with the `Seo` component and remix's own Script, Head etc...
 
Change this
```jsx
export default function App() {
  const data = useLoaderData<typeof loader>();
  const locale = data.selectedLocale ?? DEFAULT_LOCALE;
  const hasUserConsent = true;

  useAnalytics(hasUserConsent, locale);

  return (
    <html lang={locale.language}>
      <head>
        <Seo />
        <Meta />
        <Links />
      </head>
      <body>
        <Layout
          layout={data.layout as LayoutData}
          key={`${locale.language}-${locale.country}`}
        >
          <Outlet />
        </Layout>
        <ScrollRestoration />
        <Scripts />
      </body>
    </html>
  );
}
```

to

```jsx
export default function App() {
  const data = useLoaderData<typeof loader>();
  const locale = data.selectedLocale ?? DEFAULT_LOCALE;
  const hasUserConsent = true;

  return (
    <html lang={locale.language}>
      <head>
        <Seo />
        <Meta />
        <Links />
      </head>
      <body>
        <Layout
          layout={data.layout as LayoutData}
          key={`${locale.language}-${locale.country}`}
        >
          <Outlet />
        </Layout>
        <ScrollRestoration />
        <Analytics userConsent={hasUserConsent} locale={locale} />
        <Scripts />
      </body>
    </html>
  );
}
```

### HOW to test your changes?

<!--
  Give as much information for the reviewer to test your changes locally. A thorough step-by-step guide will go along-way.
-->

#### Post-merge steps

<!--
  If changes require post-merge steps, for example merging and publishing [documentation](https://shopify.dev) changes,
  specify it in this section and add the label "includes-post-merge-steps".
  If it doesn't, feel free to remove this section.
-->

#### Checklist

- [ ] I've read the [Contributing Guidelines](CONTRIBUTING.md)
- [ ] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [ ] I've added a [changeset](CONTRIBUTING.md#changesets) if this PR contains user-facing or noteworthy changes
- [ ] I've added [tests](CONTRIBUTING.md#testing) to cover my changes

<!--
 THANK YOU for your pull request! Members from the Hydrogen team will review these changes and provide feedback as soon as they are available.
-->

juanpprieto committed:
add products route ld+json and refactor analytics and seo helpers

frandiox committed:
Refresh app when hydrogen-react changes (#595)

juanpprieto committed:
change root.tsx analytics to loader pattern and add ldJson

juanpprieto committed:
cleanup unused imports

juanpprieto committed:
cleanup unused variables

Commit:
Hard code date in robots tag (#597)

Co-authored-by: Matt Seccafien <462077+cartogram@users.noreply.github.com>

wizardlyhel commented:
I'm good with the `analyticPayload` clean up but to convert `useAnalytics` hook to a component only for consistency is weird. `useAnalytics` hook doesn't need to render anything.

blittle commented:
> I'm good with the `analyticPayload` clean up but to convert `useAnalytics` hook to a component only for consistency is weird. `useAnalytics` hook doesn't need to render anything.

A component doesn't need to render anything. There's a lot of precedence in the React ecosystem to have components that don't render. I think the bigger question is the dev ergonomics.  Do we ever expect `useAnalytics` to return some sort of data? If it's a component, then you gotta use render props to pass data out. Pros and cons to both.

frehner commented:
Slight preference on the "custom hook" form over a "component that wraps a custom hook" form, but it's not strongly held. I don't see the need to make a component that does nothing, when custom hooks are a React ecosystem thing as well (and are also widely used).

blittle commented:
I think my preference also leans towards the hook, because I totally imagine that the hook could return something, and I think it's easier to adapt.

benjaminsehl commented:
If deleting the component is a shortcut to removing analytics reporting full stop -- that is kind of a nice escape hatch in an exploration of HIPAA compliance.

I don't feel really strongly either way; but I do think consistency of ergonomic is a good principle to follow -- so long as it doesn't compromise the overall DX or performance.

Likewise -- consistency here might mean we move to create a "useSeo" hook in hydrogen-react instead of a component for Analytics -- it might also mean that we say "we will consistently use components when we return markup, hooks when we don't". 

I still have to dig through the SEO proposals but I'd say we should separate the two ideas -- don't necessarily shoot down the Analytics component idea, but don't commit to merging yet either. Just let it simmer for a bit and get some dogfooding in to see how it feels. 

Separately if the SEO improvements feel great, great! (Reviewing now!)

benjaminsehl commented:
I will say -- analytics component with prop for user consent is really simple. I do like that, but need to refresh my memory with the alternative.

benjaminsehl reviewed (approved):
This DX is so much cleaner and easier to follow -- obsessed with this.

Did not top hat but approve the suggestion.

Also, glad you're using Brand Settings API.

One point of consistency -- for the brand, can we opt to use Brand API as the default across the board, and then we can fall back to the others. I believe that we can get most of what we need for base OG tags from that (need to go check reference).

benjaminsehl reviewed (commented):
_Edit: Approved by accident, meant to comment — since this is WIP and didn't top hat, can't approve._

----

If deleting the component is a shortcut to removing analytics reporting full stop -- that is kind of a nice escape hatch in an exploration of HIPAA compliance.

I don't feel really strongly either way; but I do think consistency of ergonomic is a good principle to follow -- so long as it doesn't compromise the overall DX or performance.

Likewise -- consistency here might mean we move to create a "useSeo" hook in hydrogen-react instead of a component for Analytics -- it might also mean that we say "we will consistently use components when we return markup, hooks when we don't". 

I still have to dig through the SEO proposals but I'd say we should separate the two ideas -- don't necessarily shoot down the Analytics component idea, but don't commit to merging yet either. Just let it simmer for a bit and get some dogfooding in to see how it feels. 

----

On the SEO stuff… 

This DX is so much cleaner and easier to follow -- love this direction.

Also, glad you're using Brand Settings API. We should push to see if they can expand what's exposed to include all of the social links, etc. so we can make OG defaults use Brand as the source of truth. Makes it a really nice clean place to edit a single screen and have your full app get up to date.

cartogram reviewed (approved):
🚀 @juanpprieto Thank you for helping with this!

I actually like the component that renders nothing pattern.... makes it easy to be consistent (you can move a hook into a component, but its awkward/sometimes impossible to go the the other way). So @benjaminsehl we can't do the `useSeo` hook because we produce markup (we could explore this limitation more though).

I think there has always been a good amount of sharing between `seo` and `analytics` in trying to come up with consistent APIs. Its great to see that continue to play out here ❤️

benjaminsehl commented on templates/demo-store/app/routes/($lang)/products/$productHandle.tsx:
```diff
@@ -499,6 +461,93 @@ function ProductDetail({
+  };
+}
+
+function seoPayload({
```
I think it's worth bikeshedding! Less about this specific one and more for the precedent of the pattern overall.

From a design elegance POV -- I like that it's a single function; maybe could accept an arg for the resource type.

However -- I'm being naive about in that suggestion? Is there a reason to be more imperative w/ function names? Like would it be a lot more prone to getting things wrong or something?

juanpprieto committed:
improve root and productHandle types

migrate analytic payloads helpers to lib/analytics

seo + analytics demo store implementation

cleanup analytics refactor
